File: Sigook.Web/src/types/runner.ts
// Runner types. Mirror backend Covenant.Common.Models.Runner.
// The API (System.Text.Json, no JsonStringEnumConverter) serializes enums as
// their NUMERIC value, so these must match the backend int values exactly.

export enum RunnerType {
  Active = 1,
  Passive = 2,
}

export enum RunnerStatus {
  SentToClient = 1,
  InterviewScheduled = 2,
  InterviewRescheduled = 3,
  NoLongerAvailable = 4,
  NoShow = 5,
  WaitingForInterviewFeedback = 6,
  WaitingForFinalDecision = 7,
  Rejected = 8,
  InOnboardingProcess = 9,
  Hired = 10,
}

export enum InterviewType {
  Phone = 1,
  Video = 2,
  Onsite = 3,
}

export enum InterviewStatus {
  Scheduled = 1,
  Rescheduled = 2,
}

// GetRunnersSortBy on the backend: Name=0, Status=1, Type=2, CreatedAt=3.
export enum RunnerSortBy {
  Name = 0,
  Status = 1,
  Type = 2,
  CreatedAt = 3,
}

export const RUNNER_STATUS_LABELS: Record<RunnerStatus, string> = {
  [RunnerStatus.SentToClient]: 'Sent to Client',
  [RunnerStatus.InterviewScheduled]: 'Interview scheduled',
  [RunnerStatus.InterviewRescheduled]: 'Interview rescheduled',
  [RunnerStatus.NoLongerAvailable]: 'No longer available',
  [RunnerStatus.NoShow]: 'No show',
  [RunnerStatus.WaitingForInterviewFeedback]: 'Waiting for interview feedback',
  [RunnerStatus.WaitingForFinalDecision]: 'Waiting for final decision',
  [RunnerStatus.Rejected]: 'Rejected',
  [RunnerStatus.InOnboardingProcess]: 'In onboarding process',
  [RunnerStatus.Hired]: 'Hired',
};

export const RUNNER_STATUSES: RunnerStatus[] = [
  RunnerStatus.SentToClient,
  RunnerStatus.InterviewScheduled,
  RunnerStatus.InterviewRescheduled,
  RunnerStatus.NoLongerAvailable,
  RunnerStatus.NoShow,
  RunnerStatus.WaitingForInterviewFeedback,
  RunnerStatus.WaitingForFinalDecision,
  RunnerStatus.Rejected,
  RunnerStatus.InOnboardingProcess,
  RunnerStatus.Hired,
];

export const RUNNER_TYPE_LABELS: Record<RunnerType, string> = {
  [RunnerType.Active]: 'Active',
  [RunnerType.Passive]: 'Passive',
};

export const RUNNER_TYPES: RunnerType[] = [RunnerType.Active, RunnerType.Passive];

export const INTERVIEW_TYPE_LABELS: Record<InterviewType, string> = {
  [InterviewType.Phone]: 'Phone',
  [InterviewType.Video]: 'Video',
  [InterviewType.Onsite]: 'Onsite',
};

export const INTERVIEW_TYPES: InterviewType[] = [InterviewType.Phone, InterviewType.Video, InterviewType.Onsite];

export const INTERVIEW_STATUS_LABELS: Record<InterviewStatus, string> = {
  [InterviewStatus.Scheduled]: 'Scheduled',
  [InterviewStatus.Rescheduled]: 'Rescheduled',
};

// Mirror of the backend guard Runner.CanAddInterview: interviews can only be
// added when the runner is in 'Interview scheduled' or 'Interview rescheduled'.
export function canAddInterview(status: RunnerStatus): boolean {
  return status === RunnerStatus.InterviewScheduled || status === RunnerStatus.InterviewRescheduled;
}

export interface RunnerStatusHistoryItem {
  id: string;
  previousStatus?: RunnerStatus | null;
  newStatus: RunnerStatus;
  changedBy?: string;
  changedByEmail?: string;
  changedAt: string;
  comments?: string;
}

export interface RunnerInterview {
  id: string;
  scheduledDate: string;
  type: InterviewType;
  interviewer?: string;
  status: InterviewStatus;
  feedback?: string;
  notes?: string;
  rescheduleCount: number;
  createdAt: string;
  createdBy?: string;
  rescheduledAt?: string | null;
}

// Mirrors backend RunnerListModel.
export interface RunnerListItem {
  id: string;
  numberId: number;
  agencyId: string;
  requestId: string;
  workerProfileId?: string | null;
  candidateId?: string | null;
  name: string;
  email?: string;
  type: RunnerType;
  status: RunnerStatus;
  startDate?: string | null;
  interviewsCount: number;
  createdAt: string;
  createdBy?: string;
}

// Mirrors backend RunnerDetailModel.
export interface RunnerDetail {
  id: string;
  numberId: number;
  requestId: string;
  workerProfileId?: string | null;
  candidateId?: string | null;
  name: string;
  email?: string;
  type: RunnerType;
  status: RunnerStatus;
  startDate?: string | null;
  createdAt: string;
  createdBy?: string;
  updatedAt?: string | null;
  statusHistory: RunnerStatusHistoryItem[];
  interviews: RunnerInterview[];
}

// Filter for GET .../Runners. Mirrors backend GetRunnersFilter.
export interface AgencyRunnerFilter {
  pageIndex?: number;
  pageSize?: number;
  isDescending?: boolean;
  sortBy?: RunnerSortBy;
  requestId?: string;
  name?: string;
  type?: RunnerType | null;
  statuses?: RunnerStatus[];
  createdAtFrom?: string | null;
  createdAtTo?: string | null;
}

// Body for POST .../Runners. Mirrors backend RunnerCreateModel.
export interface CreateRunnerModel {
  workerProfileId?: string | null;
  candidateId?: string | null;
  type: RunnerType;
}

// Body for PUT .../Runners/{id}/Status. Mirrors backend ChangeRunnerStatusModel.
// startDate is only used when status === RunnerStatus.Hired.
export interface ChangeRunnerStatusModel {
  status: RunnerStatus;
  comments?: string;
  startDate?: string | null;
}

// Body for POST .../Runners/{id}/Interview. Mirrors backend RunnerInterviewCreateModel.
export interface CreateRunnerInterviewModel {
  scheduledDate: string;
  type: InterviewType;
  interviewer?: string;
  notes?: string;
}

// Body for PUT .../Runners/{id}/Interview/{interviewId}/Reschedule.
export interface RescheduleRunnerInterviewModel {
  newDate: string;
}

// Mirrors backend RunnerStartingTodayModel. Hired runners (non direct-hiring)
// whose start date is within the first follow-up days, surfaced to the recruiter
// who hired them so they can confirm attendance on the punch card.
export interface RunnerStartingToday {
  runnerId: string;
  requestId: string;
  requestNumberId: number;
  jobTitle: string;
  companyName: string;
  workerProfileId: string;
  workerName: string;
  startDate: string;
  dayNumber: number;
}

File: Covenant.Api/Covenant.Common/Entities/Request/Runners/Runner.cs
using Covenant.Common.Entities.Worker;
using Covenant.Common.Enums;
using Covenant.Common.Functionals;

namespace Covenant.Common.Entities.Request.Runners;

public class Runner
{
    private Runner()
    {
    }

    public Guid Id { get; private set; } = Guid.NewGuid();
    public long NumberId { get; private set; }

    public Guid AgencyId { get; private set; }
    public Agency.Agency Agency { get; private set; }

    public Guid RequestId { get; private set; }
    public Request Request { get; private set; }

    public Guid WorkerProfileId { get; private set; }
    public WorkerProfile WorkerProfile { get; private set; }

    public Guid? RequestRecruiterId { get; private set; }
    public RequestRecruiter RequestRecruiter { get; private set; }

    public RunnerType Type { get; private set; }
    public RunnerStatus Status { get; private set; }
    public DateTime? StartDate { get; private set; }

    public DateTime CreatedAt { get; private set; } = DateTime.Now;
    public Guid CreatedBy { get; private set; }
    public DateTime? UpdatedAt { get; private set; }
    public Guid? UpdatedBy { get; private set; }

    private readonly List<RunnerStatusHistory> _statusHistory = new();
    public IEnumerable<RunnerStatusHistory> StatusHistory => _statusHistory;

    private readonly List<RunnerInterview> _interviews = new();
    public IEnumerable<RunnerInterview> Interviews => _interviews;

    public static Result<Runner> CreateFromWorker(Guid agencyId, Guid requestId, Guid workerProfileId, RunnerType type, Guid createdBy, Guid? requestRecruiterId = null)
    {
        var runner = new Runner
        {
            AgencyId = agencyId,
            RequestId = requestId,
            WorkerProfileId = workerProfileId,
            RequestRecruiterId = requestRecruiterId,
            Type = type,
            Status = RunnerStatus.SentToClient,
            CreatedBy = createdBy
        };
        runner._statusHistory.Add(RunnerStatusHistory.Create(runner.Id, null, RunnerStatus.SentToClient, createdBy, null));
        return Result.Ok(runner);
    }

    public static bool CanAddInterview(RunnerStatus status) =>
        status is RunnerStatus.InterviewScheduled or RunnerStatus.InterviewRescheduled;

    public Result ChangeStatus(RunnerStatus next, Guid changedBy, string comments = null, DateTime? startDate = null)
    {
        if (Status == RunnerStatus.Hired)
            return Result.Fail("A hired runner's status cannot be changed");
        if (next == RunnerStatus.Hired && startDate is null)
            return Result.Fail("A start date is required to hire a runner");
        var previous = Status;
        Status = next;
        UpdatedAt = DateTime.Now;
        UpdatedBy = changedBy;
        if (next == RunnerStatus.Hired) StartDate = startDate;
        _statusHistory.Add(RunnerStatusHistory.Create(Id, previous, next, changedBy, comments));
        return Result.Ok();
    }

    public Result<RunnerInterview> AddInterview(DateTime scheduledDate, InterviewType type, string interviewer, string notes, Guid createdBy)
    {
        if (!CanAddInterview(Status))
            return Result.Fail<RunnerInterview>("Interviews can only be added when the runner status is 'Interview scheduled' or 'Interview rescheduled'");
        var interview = RunnerInterview.Create(Id, scheduledDate, type, interviewer, notes, createdBy);
        _interviews.Add(interview);
        UpdatedAt = DateTime.Now;
        UpdatedBy = createdBy;
        return Result.Ok(interview);
    }

    public Result RescheduleInterview(Guid interviewId, DateTime newDate, Guid rescheduledBy)
    {
        if (!CanAddInterview(Status))
            return Result.Fail("Interviews can only be rescheduled when the runner status is 'Interview scheduled' or 'Interview rescheduled'");
        var interview = _interviews.FirstOrDefault(i => i.Id == interviewId);
        if (interview is null) return Result.Fail("Interview not found");
        var result = interview.Reschedule(newDate, rescheduledBy);
        if (!result) return result;
        UpdatedAt = DateTime.Now;
        UpdatedBy = rescheduledBy;
        if (Status != RunnerStatus.InterviewRescheduled)
            ChangeStatus(RunnerStatus.InterviewRescheduled, rescheduledBy);
        return Result.Ok();
    }
}
